Feature - initial publish raw CloudEvents

stijnmoreels opened the pull request:
Relates to #117 
Closes #123

arcus-automation commented:
A new preview package for `Arcus.EventGrid.All` is available on MyGet.

You can pull it locally via the CLI:
```shell
PM> Install-Package Arcus.EventGrid.All -Version 20200225.0.0-PR-119 -Source https://www.myget.org/F/arcus/api/v3/index.json
```

stijnmoreels commented:
Hmm, question: when we have to publish raw events as CloudEvents, what should the required **source** property be in the event?

stijnmoreels commented:
I changed the publisher to publish all the kind of events as CloudEvents since that's the only correct way I guess. We should I think wait for  #112 to be completed before we can publish custom events as cloud events .

tomkerkhove commented on src/Arcus.EventGrid.Publishing/CloudEvents/CloudEventBatchContent.cs:
```diff
@@ -0,0 +1,97 @@
+    /// <summary>
+    /// Represents a <see cref="HttpContent"/> model containing a batch of <see cref="CloudEvent"/>s instances.
+    /// </summary>
+    public class CloudEventBatchContent : HttpContent
```
Why are we re-inventing the wheel instead of building on top of the SDK?https://github.com/cloudevents/sdk-csharp#http---systemnethttphttpclient

stijnmoreels replied:
Not re-inventing, but building on top. The SDK doesn't support batched cloud events, so made this to support the current `HttpContent` implementation for cloud events.

tomkerkhove replied:
Makes sense, thanks for letting us know!

stijnmoreels replied:
Added issue for this: https://github.com/cloudevents/sdk-csharp/issues/44

tomkerkhove commented on src/Arcus.EventGrid.Publishing/EventGridPublisher.cs:
```diff
@@ -66,9 +66,10 @@ internal EventGridPublisher(Uri topicEndpoint, string authenticationKey, Policy
         /// <param name="eventBody">Body of the event</param>
-        public async Task PublishRawAsync(string eventId, string eventType, string eventBody)
+        /// <param name="eventSchema">The schema in which the event should be published.</param>
+        public async Task PublishRawAsync(string eventId, string eventType, string eventBody, EventSchema eventSchema = EventSchema.EventGrid)
```
Do we use that or `PublishRawEventGridEventAsync` and `PublishRawCloudEventAsync`?

stijnmoreels replied:
I think the API on this would be even more complex if we would do that.

tomkerkhove replied:
Can you give an example please? (other than the event source)

stijnmoreels replied:
Don't know what you mean.
What I meant was that if we introduce another type in the signature, it would become more complex.

Now, we already have `PublishRaw/Many`, `Publish/Many (custom)`, `Publish/Many(cloud)`.

tomkerkhove replied:
My suggestion would remove the current publish raw, not the others

stijnmoreels replied:
Hmm, question, what do we do with the `PublishRawAsync(RawEvent)` ones? They also have to be converted to cloud events but don't have a required `source`. 😞 .

tomkerkhove replied:
That's what we've discussed yesterday where I would remove PublishRawAsync and replace it with PublishRawAsEventGridAsync & PublishRawAsCloudEventAsync so you can add required parameters

stijnmoreels replied:
Hmm, I can't even get the existing integration tests to work locally 😭 . A little hard to add some CloudEvent tests as well.

arcus-automation commented:
A new preview package for `Arcus.EventGrid.All` is available on MyGet.

You can pull it locally via the CLI:
```shell
PM> Install-Package Arcus.EventGrid.All -Version 20200228.0.0-PR-119 -Source https://www.myget.org/F/arcus/api/v3/index.json
```

tomkerkhove commented on src/Arcus.EventGrid.Publishing/EventGridPublisher.cs:
```diff
@@ -56,31 +58,31 @@ internal EventGridPublisher(Uri topicEndpoint, string authenticationKey, Policy
+        /// <param name="eventSubject">The subject of the event.</param>
+        /// <param name="dataVersion">The data version of the event body.</param>
+        /// <param name="eventTime">The time when the event occured.</param>
+        public async Task PublishRawEventGridAsync(string eventId, string eventType, string eventBody, string eventSubject, string dataVersion, DateTimeOffset eventTime)
```
```suggestion
        public async Task PublishRawEventGridEventAsync(string eventId, string eventType, string eventBody, string eventSubject, string dataVersion, DateTimeOffset eventTime)
```

stijnmoreels replied:
Ok

tomkerkhove commented on src/Arcus.EventGrid.Publishing/EventGridPublisher.cs:
```diff
@@ -56,31 +58,31 @@ internal EventGridPublisher(Uri topicEndpoint, string authenticationKey, Policy
+        /// <param name="eventId">The unique identifier of the event.</param>
+        /// <param name="eventType">The type of the event.</param>
+        /// <param name="eventBody">The body of the event.</param>
+        public async Task PublishRawEventGridAsync(string eventId, string eventType, string eventBody)
```
```suggestion
        public async Task PublishRawEventGridEventAsync(string eventId, string eventType, string eventBody)
```

stijnmoreels replied:
Ok

stijnmoreels commented:
> LGTM but we'll need CloudEvent tests as well, would you mind adding some please?e

Yes, I'm gonna need some help with that. Infrastructure-stuff and all. I even can't get the existing integration tests to work locally 😭 .

tomkerkhove commented:
What do you need and what is it that you can't create?

arcus-automation commented:
A new preview package for `Arcus.EventGrid.All` is available on MyGet.

You can pull it locally via the CLI:
```shell
PM> Install-Package Arcus.EventGrid.All -Version 20200302.0.0-PR-119 -Source https://www.myget.org/F/arcus/api/v3/index.json
```

stijnmoreels commented:
> What do you need and what is it that you can't create?

I need a CloudEvent infrastructure like for the EventGrid events. But I can't to seem to alter the custom deployment template to make sure the input schema is CloudEvent instead of EventGrid events.

arcus-automation commented:
A new preview package for `Arcus.EventGrid.All` is available on MyGet.

You can pull it locally via the CLI:
```shell
PM> Install-Package Arcus.EventGrid.All -Version 20200303.0.0-PR-119 -Source https://www.myget.org/F/arcus/api/v3/index.json
```

tomkerkhove commented:
We'll need both but feel free to manually just create one in your subscription and if everything is up and running I can do the required part in the Arcus subscription

stijnmoreels commented:
> We'll need both but feel free to manually just create one in your subscription and if everything is up and running I can do the required part in the Arcus subscription

I know, but how do you change the deployment template to use the CloudEvent as input schema?

tomkerkhove commented:
Just create it via the portal :)

Alternatively you can just check https://github.com/arcus-azure/arcus.eventgrid/pull/116

stijnmoreels commented:
Ok, I added some integration tests for the CloudEvents and I added also some appsettings keys to the devops lirbary so they work here. It still uses resources on my local subscription but that's fine I guess until #115 is resolved.

arcus-automation commented:
A new preview package for `Arcus.EventGrid.All` is available on MyGet.

You can pull it locally via the CLI:
```shell
PM> Install-Package Arcus.EventGrid.All -Version 20200304.0.0-PR-119 -Source https://www.myget.org/F/arcus/api/v3/index.json
```

tomkerkhove commented:
It depends on the current setup - So you now use an event grid topic with Event Grid schema and one with CloudEvents, right?

#115 is only for setting up the testing infrastructure for using it integration tests, not specific for our testing.

stijnmoreels commented:
> It depends on the current setup - So you now use an event grid topic with Event Grid schema and one with CloudEvents, right?

Yes, right. Both on my subscription though. I'll leave it up to you to update the stuff with resources on the Arcus resource group (with the right names and all).

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/Arcus.EventGrid.Tests.Integration.csproj:
```diff
@@ -1,10 +1,11 @@
 
   <PropertyGroup>
-    <TargetFrameworks>netcoreapp2.0;netcoreapp3.0</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.2;netcoreapp3.0</TargetFrameworks>
```
Are our libs, other than testing, using .NET Standard? If so, fine

stijnmoreels replied:
The Arcus.EventGridSecurity is `<TargetFrameworks>netstandard2.0;netcoreapp3.0</TargetFrameworks>`

tomkerkhove replied:
👌

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/appsettings.json:
```diff
@@ -1,12 +1,24 @@
+      },
+      "ServiceBus": {
+        "TopicName": "#{Arcus_EventGridEvent_ServiceBus_TopicName}#",
+        "ConnectionString": "#{Arcus_EventGridEvent_ServiceBus_ConnectionString}#"
```
Can't we move this to `Arcus:ServiceBus:ConnectionString`?

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/appsettings.json:
```diff
@@ -1,12 +1,24 @@
-    "EventGrid": {
-      "TopicEndpoint": "#{Arcus_EventGrid_TopicEndpoint}#",
-      "EndpointKey": "#{Arcus_EventGrid_EndpointKey}#"
+    "EventGridEvent": {
```
Please align tokens to use our existing naming convention

tomkerkhove replied:
Can you also add them to the build please? I don't see our build yaml being updated

stijnmoreels replied:
No, I updated the library in azure devops.

tomkerkhove replied:
Ok but can we at least align the tokenization naming to align with others?

stijnmoreels replied:
Have actually no idea what you're talking about, or what you want to change 😄 .

tomkerkhove replied:
`Arcus_EventGrid_EndpointKey` -> `EventGrid.EndpointKey`

tomkerkhove replied:
- Don't forget to update the variable group as all your tests are broken now
- Don't forget to remove the `Arcus.` prefix similar to our existing approach

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/Publishing/EventGridEventPublishingTests.cs:
```diff
@@ -4,164 +4,102 @@
         public async Task PublishSingleRawEventWithDetailedInfo_WithBuilder_ValidParameters_Succeeds()
         {
             // Arrange
-            var topicEndpoint = Configuration.GetValue<string>("Arcus:EventGrid:TopicEndpoint");
```
Why was this test removed?

stijnmoreels replied:
We can't publish `RawEvent`'s anymore, like we discussed.

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/Publishing/CloudEventPublishing.cs:
```diff
@@ -0,0 +1,196 @@
+                cloudEvent.Source,
+                cloudEvent.Subject,
+                rawEventBody,
+                cloudEvent.Time ?? default(DateTimeOffset));
```
We control this so why do we need to check this?

stijnmoreels replied:
Bc `CloudEvent.Time` is a `Nullable<DateTime>`.

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/Publishing/CloudEventPublishing.cs:
```diff
@@ -0,0 +1,196 @@
+            _testOutput.WriteLine($"Event '{eventId}' published - {JsonConvert.SerializeObject(events)}");
+        }
+
+        private static void AssertReceivedNewCarRegisteredEvent(string eventId, string eventType, string eventSubject, string licensePlate, string receivedEvent)
```
Shouldn't we share this with other test? Maybe add a custom `Assert.ReceivedNewCarRegisteredEvent`

stijnmoreels replied:
Yes, was also thinking that. But I'm also not sure how the testing infrastructure will eventually look like. Like input/output cloudevent/evengridevent schema. So... not sure if we can already fully make this allign.
But I can move it for now.

tomkerkhove replied:
What schema we use to wrap our payload in will not impact how we assert our payload itself, no?

stijnmoreels replied:
No, I changed that too. It uses the general `Event` to assert.

tomkerkhove replied:
LGTM

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/appsettings.json:
```diff
@@ -1,12 +1,24 @@
 {
```
So what I'm seeing is that you have:
- One Service Bus namespace with 2 topics
- One Event Grid topic with Event Grid schema
- One Event Grid topic with CloudEvents schema

If so, I'll add an issue to reflect this in our landscape

stijnmoreels replied:
I have two ServiceBus namespaces. I actually just deployed the test instracture twice but the second time with a CloudEvent input schema.

tomkerkhove replied:
Yeah we should consolidate those :)

tomkerkhove replied:
LGTM

stijnmoreels replied:
Ok, but definitly in another PR, since it's not even the final Azure resources.

stijnmoreels replied:
> Feel free to create an issue for the Azure changes and assign it to me

Done: https://github.com/arcus-azure/arcus.eventgrid/issues/127

tomkerkhove commented on src/Arcus.EventGrid/Parsers/EventParser.cs:
```diff
@@ -107,5 +106,11 @@ internal static Event ParseJObject(JObject rawInput)
             }
         }
+
+        private static bool IsCloudEvent(JObject jObject)
```
Idea - Provide unit tests for this or is it covered somewhere?

stijnmoreels replied:
It's actually from the SDK.

tomkerkhove replied:
Even then, shouldn't we provide testing as it's pretty crucial?

For example, do we have unit tests where we have all 4 flavors (CE 0.1, CE 0.2, CE 1.0 & EG) and verify they are parsed correctly?

tomkerkhove replied:
Seperate PR is ok

stijnmoreels replied:
The EventGrid only allows the v1.0 CloudEvent schema, so don't even sure if we should allow others.

tomkerkhove replied:
Existing endpoints before 1.0 still support 0.1 & 0.2

stijnmoreels replied:
#126

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/Fixture/EventGridTopicEndpoint.cs:
```diff
@@ -0,0 +1,108 @@
+        /// Creates a <see cref="EventGridTopicEndpoint"/> implementation that uses CloudEvent's as input event schema.
+        /// </summary>
+        /// <param name="testOutput">The test logger to write diagnostic messages during the availability of the endpoint.</param>
+        public static async Task<EventGridTopicEndpoint> CreateForCloudEventAsync(ITestOutputHelper testOutput)
```
Aren't we mixing creating the endpoint publisher and consuming the events in one class?

tomkerkhove replied:
LGTM

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/Publishing/CloudEventPublishing.cs:
```diff
@@ -0,0 +1,196 @@
+        /// </summary>
+        public async Task InitializeAsync()
+        {
+            _endpoint = await EventGridTopicEndpoint.CreateForCloudEventAsync(_testOutput);
```
See EventGridTopicEndpoint, I'd argue this class does 2 things:
- Provide an endpoint to create publisher for (which we do for every test via BuildPublisher, while we could only do it once? 🤔 )
- Provide host to receive events on

stijnmoreels replied:
Ah, yes, we could maybe move the creation logic of the event publisher to somehwere else.

tomkerkhove replied:
LGTM

stijnmoreels commented:
So... @tomkerkhove , only the ARM templates and the updates in the Azure resource groups are needed to delegate the tests against the R&D subscription. Definitely other PR I guess, so this isn't blocked.

tomkerkhove commented:
Feel free to create an issue for the Azure changes and assign it to me

tomkerkhove commented on src/Arcus.EventGrid.Tests.Integration/appsettings.json:
```diff
@@ -1,12 +1,24 @@
+    "CloudEvent": {
+      "EventGrid": {
+        "TopicEndpoint": "#{CloudEvent.EventGrid.TopicEndpoint}#",
+        "EndpointKey": "#{CloudEvent.EventGrid.EndpointKey}#"
```
@stijnmoreels  I've noticed these are still specific per schema type, can you move this above so that connection string is shared please? (topic will be specific per schema though)

stijnmoreels replied:
So specific for eventgrid topic but same service bus connection?
Will the tests then still work?

tomkerkhove replied:
It depends how you've configured it but they should share the same Service Bus namespace.

But maybe it's better that I'll pick this up with the Azure infra changes?

stijnmoreels replied:
Yes, I think so too 😞 . Thanks!!
